File: Labs/Lab 6/Lab6.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Lab6 {
	public static boolean canFinish(int numCourses, int[][] prerequisites) {
		Map<Integer, List<Integer>> graph = new HashMap<>();

        for (int[] prerequisite : prerequisites) {
            int from = prerequisite[1];
            int to = prerequisite[0];
            if (!graph.containsKey(to)) {
                graph.put(to, new ArrayList<>());
            }
            graph.get(to).add(from);
        }

        boolean[] visited = new boolean[numCourses];
        for (int i = 0; i < numCourses; i++) {
            if (!visited[i] && hasCycle(graph, i, visited)) {
                return false;
            }
        }

        return true;
    }

    private static boolean hasCycle(Map<Integer, List<Integer>> graph, int i, boolean[] visited) {
        if (visited[i]) return true; 
        visited[i] = true;

        List<Integer> neighbors = graph.get(i);
        if (neighbors == null) {
            neighbors = new ArrayList<>();
            graph.put(i, neighbors);
        }
        for (int neighbor : neighbors) {
            if (hasCycle(graph, neighbor, visited)) {
                return true; 
            }
        }
        visited[i] = false; 
        return false;
    }

    public static void main(String[] args) {
        int numCourses = 2;
        int[][] prerequisites = {{1, 0}, {0, 1}};
        System.out.println(canFinish(numCourses, prerequisites)); 
        

        int numCourse = 2;
        int[][] prerequisite = {{1, 0}};
        System.out.println(canFinish(numCourse, prerequisite)); 
    }
}

File: Labs/Lab 6/Lab 6.py
def can_finish(num_courses, prerequisites):
  graph = {}

  for prerequisite in prerequisites:
      to, from_ = prerequisite
      if to not in graph:
          graph[to] = []
      graph[to].append(from_)

  visited = [False] * num_courses
  for i in range(num_courses):
      if not visited[i] and has_cycle(graph, i, visited):
          return False

  return True

def has_cycle(graph, i, visited):
  if visited[i]:
      return True
  visited[i] = True

  neighbors = graph.get(i, [])
  if not neighbors:
      graph[i] = []
  for neighbor in neighbors:
      if has_cycle(graph, neighbor, visited):
          return True

  visited[i] = False
  return False

# Test cases
num_courses = 2
prerequisites = [[1, 0], [0, 1]]
print(can_finish(num_courses, prerequisites)) 

num_courses = 2
prerequisites = [[1, 0]]
print(can_finish(num_courses, prerequisites))
